lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,232 @@
+      end
+      DiscoverableResource.const_get(kind)
+    end
+
+    def self.with_retries(retries = 3, backoff = 10)
+      yield
+    rescue KubeException => err
+      if (retries -= 1) > 0
+        logger.warn("Retrying to discover CustomResourceDefinitions: #{err}")
+        sleep(backoff)
+        retry
+      else
+        logger.warn("Unable to discover CustomResourceDefinitions: #{err}")
+      end
+    end
+
+    def self.discover_crd(logger:, client:)
+      @child_classes = Set.new
+      resources = client.get_custom_resource_definitions
+      resources.each do |res|
+        kind = res.spec.names.kind
+        # Remove and redefine the class if it already exists so we can be up to date.
+        if DiscoverableResource.const_defined?(kind)
+          klass = DiscoverableResource.const_get(kind)
+          DiscoverableResource.send(:remove_const, kind)
+          child_classes.delete(klass)
+        end
+        generate_resource(group: res.spec.group,
+                           version: res.spec.version,
+                           kind: kind,
+                           annotations: res.metadata.annotations)
+        logger.info("  - #{res.spec.group}/#{res.spec.version}/#{kind}")
+      end
+    end
+
+    def self.generate_resource(group:, version:, kind:, annotations:)
+      deploy_metadata = annotations[DEPLOY_METADATA_ANNOTATION] || '{}'
+      metadata = JSON.parse(deploy_metadata)
+      unless metadata.is_a?(Hash)
+        raise FatalDeploymentError, "Invalid metadata for #{kind} #{metadata.inspect}"
+      end
+
+      predeploy_dependencies = [] << metadata['predeploy-dependencies']
+      predeploy_dependencies.flatten!
+      predeploy_dependencies.compact!
+
+      prunable = parse_bool(metadata['prunable'])
+      predeploy = parse_bool(metadata['predeploy'])
+      status_query = metadata['status-query']
+      success_status = metadata['status-success']
+
+      resource_template = ERB.new <<-CLASS
+        class #{kind.capitalize} < DiscoverableResource
+          GROUP = '#{group}'
+          VERSION = '#{version}'
+          PREDEPLOY = #{predeploy}
+          PRUNABLE = #{prunable}
+
+          <% unless predeploy_dependencies.empty? %>
+          PREDEPLOY_DEPENDENCIES = #{predeploy_dependencies}
+          <% end %>
+
+          <% if status_query && success_status %>
+          def deploy_succeeded?
+            getter = "get_#{kind.downcase}"
+            @client ||= DiscoverableResource.kubeclient(context: @context, resource_class: self.class)
+            raw_json = @client.send(getter, @name, @namespace, as: :raw)
+            result = query_json(query: STATUS_QUERY, json: raw_json)
+            result == SUCCESS_STATUS
+          end
+          <% end %>
+
+          self
+        end
+      CLASS
+
+      rendered_template = resource_template.result(binding)
+      klass = class_eval(rendered_template)
+      klass.const_set(:STATUS_QUERY, status_query)
```

Comment:
I've adopted the `class_eval` approach suggested by @KnVerey, but I did not want to include the `status_query` and `status_success` fields in the `class_eval` input because they can contain arbitrary user-provided data, possibly executable Ruby code. While I do not think there is particular concern for these fields to be used as attack vectors, it doesn't seem prudent to allow user data to be interpreted as arbitrary Ruby code.